File: content/projectilewars/panorama/scripts/custom_game/hotkeys.ts
let abilityPanels:Panel[] = []

function SetUpHotKeys() : void {
  $.Schedule(0.01,function(){SetUpHotKeys();})
  let inputs: [string,string,string,string,string,string,string];
  inputs = [
    "_",
    Game.GetKeybindForAbility(0),
    Game.GetKeybindForAbility(1),
    Game.GetKeybindForAbility(2),
    Game.GetKeybindForAbility(3),
    Game.GetKeybindForAbility(4),
    Game.GetKeybindForAbility(5),
  ]
  

  //if (DotaAbilityList) {
    let abName:AbilityImage ;
    let ability:number;
    let key: string;
    let pan:Panel;
    if (!Players.GetSelectedEntities(Players.GetLocalPlayer())) {
        return;
    }
    let hero = Players.GetSelectedEntities(Players.GetLocalPlayer())[0]
    if (!hero) {return}
    //let hero = Players.GetPlayerHeroEntityIndex(Players.GetLocalPlayer());
    let j = 0;
    let abilities:number = 0

    let hbar = $.GetContextPanel().GetParent().GetParent().GetParent().FindChildTraverse("BottomHealthbar")
    hbar.style.width = Entities.GetHealthPercent(hero)*7 + "px"
    
    let red = Math.round(255 - (Entities.GetHealthPercent(hero) * 2.55))
    let green =  Math.round(Entities.GetHealthPercent(hero) * 2.55)
    
    hbar.style.backgroundColor = "gradient( linear, 0% 33%, 66% 100%, from( rgb("+red*0.5+","+green*0.5+",0) ), to( rgb("+red+","+green+",0) ) )";

    for(let i = 0 ; i <=6 ; i++) {
      ability = Entities.GetAbility(hero,i)
      if (ability && Abilities.GetAbilityName(ability) != "") {
        pan = DotaAbilityList.FindChildTraverse("Ability"+(i-j  ));
        if (!abilityPanels[i]) {
          abilityPanels[i] = pan
        }

        if (pan.style.visibility != "visible") {
          pan.style.visibility = "visible"
        }
        key = inputs[i];

        pan.style.width = "100px";
        let textP = pan.FindChildTraverse("HotkeyText") as LabelPanel
        textP.text = key;
        if (pan.FindChildTraverse("AbilityLevelContainer").style.visibility != "collapse") {
          pan.FindChildTraverse("AbilityLevelContainer").style.visibility = "collapse";
        }
        if (pan.FindChildTraverse("ButtonWell").style.width != "100px") {
          pan.FindChildTraverse("ButtonWell").style.width = "100px";
          pan.FindChildTraverse("ButtonWell").style.height = "100px";
          pan.FindChildTraverse("ButtonSize").style.width = "100px";
          pan.FindChildTraverse("ButtonSize").style.height = "100px";
          pan.FindChildTraverse("Hotkey").style.visibility = "visible";
          pan.FindChildTraverse("HotkeyContainer").style.marginTop = "13%";
          pan.FindChildTraverse("HotkeyContainer").style.marginLeft = "10%";

          let numPanel = pan.FindChildTraverse("GoldCostBG")
          numPanel.style.width = "50px"
          numPanel.style.height = "50px"
          let numPanelText = pan.FindChildTraverse("GoldCost") as LabelPanel
          if (numPanelText.text && Number(numPanelText.text) > 0) {
            numPanelText.style.fontSize = "30px"
          }
        }
        
        if (i == 0) {
          if (Entities.IsDisarmed(hero) || Entities.IsRooted(hero)) {
            if (!pan.FindChildTraverse("DisarmedAbility")) {
              let disarmPanel = $.CreatePanel("Image",pan.FindChildTraverse("AbilityButton"),"DisarmedAbility") as ImagePanel
              disarmPanel.AddClass("DisarmAbility")
              // @ts-ignore
              disarmPanel.SetImage("file://{images}/custom_game/disarm.png")
              disarmPanel.style.height = "110%"
              disarmPanel.style.width = "110%"
              disarmPanel.style.verticalAlign = "center"
            }
          } else if (pan.FindChildTraverse("DisarmedAbility")) {
            $.Msg("Removed")
            pan.FindChildTraverse("DisarmedAbility").DeleteAsync(0)
          }
        }

        abilities = abilities +1 
      } else { 
        j = j + 1
      }
    }
    // Hide unused ability panels
    for (let i=6;i >= abilities;i--) {
      pan = DotaAbilityList.FindChildTraverse("Ability"+(i));
      if (pan) {
        pan.style.visibility = "collapse";
      }
    }

    /*
    for (let i=2;i < 6;i++) {
      pan = DotaAbilityList.FindChildTraverse("Ability"+(i));
      if (pan) {
        let image:AbilityImage = pan.FindChildTraverse("AbilityImage");
        let s:string = image.abilityname;
        if (s == "") {
          pan.style.visibility = "collapse";
        }
      }
    }*/
  //} 
}

function MouseTooltipManager():void {
  $.Schedule(0.01,function(){MouseTooltipManager();})
  
  if (Game.GetState() < DOTA_GameState.DOTA_GAMERULES_STATE_PRE_GAME) { 
    return
  } 
  // Abilities
  for (let panel of abilityPanels) {
    //$.Msg(panel)
    if (panel.visible == true) {
      let x = Math.abs(panel.GetPositionWithinWindow().x+35 -GameUI.GetCursorPosition()[0])
      let y = Math.abs(panel.GetPositionWithinWindow().y+70-GameUI.GetCursorPosition()[1])
      if (x < 35 && y < 35) {
        let name = panel.id.substr(7)
        let hero = Players.GetPlayerHeroEntityIndex(Players.GetLocalPlayer());
        let ability = Entities.GetAbility(hero,Number(name))
        name = Abilities.GetAbilityName(ability)
        
        $.DispatchEvent( "DOTAShowAbilityTooltip", panel,name );
        return
      } else {
        //$.DispatchEvent( "DOTAHideAbilityTooltip");
      }
    }
  }
  $.DispatchEvent( "DOTAHideAbilityTooltip");
  
}

function SilenceAbilities(kv:{silenceIndex:number[]}) {
  for (let i of kv.silenceIndex) {
    if (DotaAbilityList.FindChildTraverse("Ability"+i)) {
      let silencePanel = $.CreatePanel("Image",DotaAbilityList.FindChildTraverse("Ability"+i).FindChildTraverse("AbilityButton"),"SilencedAbility")
      silencePanel.AddClass("SilenceAbility")
      //@ts-ignore
      silencePanel.SetImage("file://{images}/custom_game/silence.png")
    }

  }
}

function UnSilenceAbilities(kv) {
  for (let i of kv.silenceIndex) {
    let pan = DotaAbilityList.FindChildTraverse("Ability"+i)

    if (pan) {
      let panel = pan.FindChildTraverse("SilencedAbility")
      if (panel) {
        panel.DeleteAsync(0)
      }
    }
  }
}

const DotaAbilityList = $.GetContextPanel().GetParent().GetParent().FindChildTraverse("abilities");
SetUpHotKeys();
MouseTooltipManager();


GameEvents.Subscribe("hero_silence_created",SilenceAbilities)
GameEvents.Subscribe("hero_silence_removed",UnSilenceAbilities)

File: content/projectilewars/panorama/scripts/custom_game/hotkeys.lua
require("typescript_lualib")
local abilityPanels = {}

function SetUpHotKeys()
    DollarStatic.Schedule($,0.01,function()
        SetUpHotKeys()
    end
)
    local inputs = nil

    inputs={"_",CScriptBindingPR_Game.GetKeybindForAbility(Game,0),CScriptBindingPR_Game.GetKeybindForAbility(Game,1),CScriptBindingPR_Game.GetKeybindForAbility(Game,2),CScriptBindingPR_Game.GetKeybindForAbility(Game,3),CScriptBindingPR_Game.GetKeybindForAbility(Game,4),CScriptBindingPR_Game.GetKeybindForAbility(Game,5)}
    local abName = nil

    local ability = nil

    local key = nil

    local pan = nil

    local hero = CScriptBindingPR_Players.GetSelectedEntities(Players,CScriptBindingPR_Players.GetLocalPlayer(Players))[0+1]

    if not hero then
        return
    end
    local j = 0

    local abilities = 0

    local hbar = Panel.FindChildTraverse(Panel.GetParent(Panel.GetParent(Panel.GetParent(DollarStatic.GetContextPanel($)))),"BottomHealthbar")

    hbar.style.width=(CScriptBindingPR_Entities.GetHealthPercent(Entities,hero)*7).."px"
    local red = math.round(255-(CScriptBindingPR_Entities.GetHealthPercent(Entities,hero)*2.55))

    local green = math.round(CScriptBindingPR_Entities.GetHealthPercent(Entities,hero)*2.55)

    hbar.style.backgroundColor="gradient( linear, 0% 33%, 66% 100%, from( rgb("..(red*0.5)..","..(green*0.5)..",0) ), to( rgb("..red..","..green..",0) ) )"
    for i=0,6,1 do
        ability=CScriptBindingPR_Entities.GetAbility(Entities,hero,i)
        if ability and (CScriptBindingPR_Abilities.GetAbilityName(Abilities,ability)~="") then
            pan=Panel.FindChildTraverse(DotaAbilityList,"Ability"..(i-j))
            if not abilityPanels[i+1] then
                abilityPanels[i+1]=pan
            end
            if pan.style.visibility~="visible" then
                pan.style.visibility="visible"
            end
            key=inputs[i+1]
            pan.style.width="100px"
            local textP = Panel.FindChildTraverse(pan,"HotkeyText")

            textP.text=key
            if Panel.FindChildTraverse(pan,"AbilityLevelContainer").style.visibility~="collapse" then
                Panel.FindChildTraverse(pan,"AbilityLevelContainer").style.visibility="collapse"
            end
            if Panel.FindChildTraverse(pan,"ButtonWell").style.width~="100px" then
                Panel.FindChildTraverse(pan,"ButtonWell").style.width="100px"
                Panel.FindChildTraverse(pan,"ButtonWell").style.height="100px"
                Panel.FindChildTraverse(pan,"ButtonSize").style.width="100px"
                Panel.FindChildTraverse(pan,"ButtonSize").style.height="100px"
                Panel.FindChildTraverse(pan,"Hotkey").style.visibility="visible"
                Panel.FindChildTraverse(pan,"HotkeyContainer").style.marginTop="13%"
                Panel.FindChildTraverse(pan,"HotkeyContainer").style.marginLeft="10%"
                local numPanel = Panel.FindChildTraverse(pan,"GoldCostBG")

                numPanel.style.width="50px"
                numPanel.style.height="50px"
                local numPanelText = Panel.FindChildTraverse(pan,"GoldCost")

                if numPanelText.text and (Number(numPanelText.text)>0) then
                    numPanelText.style.fontSize="30px"
                end
            end
            abilities=(abilities+1)
        else
            j=(j+1)
        end
    end
    for i=6,abilities,-1 do
        pan=Panel.FindChildTraverse(DotaAbilityList,"Ability"..(i))
        if pan then
            pan.style.visibility="collapse"
        end
    end
end
function MouseTooltipManager()
    DollarStatic.DispatchEvent($,"DOTAHideAbilityTooltip")
end
local DotaAbilityList = Panel.FindChildTraverse(Panel.GetParent(Panel.GetParent(DollarStatic.GetContextPanel($))),"abilities")

SetUpHotKeys()
MouseTooltipManager()
function GetBuffAbility(textureName)
    if textureName=="kobold_taskmaster_speed_aura" then
        return "item_rune_speedshot"
    else
        if textureName=="medusa_split_shot" then
            return "item_rune_multishot"
        else
            if textureName=="rune_haste" then
                return "item_rune_haste"
            else
                if textureName=="obsidian_destroyer_essence_aura" then
                    return "item_rune_castpoint"
                else
                    if textureName=="wisp_spirits" then
                        return "item_rune_turnrate"
                    end
                end
            end
        end
    end
    return ""
end
